server/src/main/java/org/apache/kafka/server/ClientMetricsManager.java

```diff
@@ -433,4 +462,35 @@ public Map<String, Pattern> matchPattern() {
             return matchPattern;
         }
     }
+
+    private final class ExpirationTimerTask extends TimerTask {
+
+        private static final long CACHE_ERROR_LOG_INTERVAL_MS = 5 * 60 * 1000; // 5 minutes
+
+        private final Uuid uuid;
+
+        private ExpirationTimerTask(Uuid uuid, long delayMs) {
+            super(delayMs);
+            this.uuid = uuid;
+        }
+
+        @Override
+        public void run() {
+            log.trace("Expiration timer task run for client instance id: {}, after delay ms: {}", uuid, delayMs);
+            if (!clientInstanceCache.remove(uuid)) {
+                /*
+                 This can only happen if the client instance is removed from the cache by the LRU
+                 eviction policy before the expiration timer task is executed. Log a warning as broker
+                 cache is not able to hold all the client instances. Log only once every CACHE_ERROR_LOG_INTERVAL_MS
+                 to avoid flooding the logs.
+                */
+                long lastErrorMs = lastCacheErrorLogMs.get();
+                if (time.milliseconds() - lastErrorMs > CACHE_ERROR_LOG_INTERVAL_MS &&
+                    lastCacheErrorLogMs.compareAndSet(lastErrorMs, time.milliseconds())) {
```

Comment:
I would log the line only if `compareAndSet` returned `true`. This means this thread won the race.

Reply:
I agree, and if I am not wrong then that's what happen with above code. Am I missing something?

Reply:
Well, if you are sure there's no race, I would use `getAndSet`. And if there is the possibility that another thread wins the race, I would just write: 

```
if  (lastCacheErrorLogMs.compareAndSet(lastErrorMs, time.milliseconds()))) {
   log.warn......
```